src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -0,0 +1,672 @@
+		}
+	}
+
+	/**
+	 * Registers the `@wordpress/interactivity` script modules.
+	 *
+	 * @since 6.5.0
+	 */
+	public function register_script_modules() {
+		wp_register_script_module(
+			'@wordpress/interactivity',
+			includes_url( 'js/dist/interactivity.min.js' )
+		);
+
+		wp_register_script_module(
+			'@wordpress/interactivity-router',
+			includes_url( 'js/dist/interactivity-router.min.js' ),
+			array( '@wordpress/interactivity' )
+		);
+	}
+
+	/**
+	 * Adds the necessary hooks for the Interactivity API.
+	 *
+	 * @since 6.5.0
+	 */
+	public function add_hooks() {
+		add_action( 'wp_footer', array( $this, 'print_client_interactivity_data' ) );
+	}
+
+	/**
+	 * Processes the interactivity directives contained within the HTML content
+	 * and updates the markup accordingly.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $html The HTML content to process.
+	 * @return string The processed HTML content. It returns the original content when the HTML contains unbalanced tags.
+	 */
+	public function process_directives( string $html ): string {
+		$p               = new WP_Interactivity_API_Directives_Processor( $html );
+		$tag_stack       = array();
+		$namespace_stack = array();
+		$context_stack   = array();
+		$unbalanced      = false;
+
+		$directive_processor_prefixes          = array_keys( self::$directive_processors );
+		$directive_processor_prefixes_reversed = array_reverse( $directive_processor_prefixes );
+
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) && false === $unbalanced ) {
+			$tag_name = $p->get_tag();
+
+			if ( $p->is_tag_closer() ) {
+				list( $opening_tag_name, $directives_prefixes ) = end( $tag_stack );
+
+				if ( 0 === count( $tag_stack ) || $opening_tag_name !== $tag_name ) {
+
+					/*
+						* If the tag stack is empty or the matching opening tag is not the
+						* same than the closing tag, it means the HTML is unbalanced and it
+						* stops processing it.
+						*/
+					$unbalanced = true;
+					continue;
+				} else {
+
+					/*
+						* It removes the last tag from the stack.
+						*/
+					array_pop( $tag_stack );
+
+					/*
+						* If the matching opening tag didn't have any directives, it can skip
+						* the processing.
+						*/
+					if ( 0 === count( $directives_prefixes ) ) {
+						continue;
+					}
+				}
+			} else {
+				$directives_prefixes = array();
+
+				foreach ( $p->get_attribute_names_with_prefix( 'data-wp-' ) as $attribute_name ) {
+
+					/*
+						* Extracts the directive prefix to see if there is a server directive
+						* processor registered for that directive.
+						*/
+					list( $directive_prefix ) = $this->extract_prefix_and_suffix( $attribute_name );
+					if ( array_key_exists( $directive_prefix, self::$directive_processors ) ) {
+						$directives_prefixes[] = $directive_prefix;
+					}
+				}
+
+				/*
+					* If this is not a void element, it adds it to the tag stack so it can
+					* process its closing tag and check for unbalanced tags.
+					*/
```

Comment:
```suggestion
				/*
				 * If this is not a void element, it adds it to the tag stack so it can
				 * process its closing tag and check for unbalanced tags.
				 */
```